File: 剑指offer/29.顺时针打印矩阵.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []

        left = 0
        right = len(matrix[0]) - 1
        top = 0
        button = len(matrix) - 1

        ans = []
        while True:
            for i in range(left, right + 1):
                ans.append(matrix[top][i])
            top += 1
            if top > button:
                break

            for i in range(top, button + 1):
                ans.append(matrix[i][right])
            right -= 1
            if left > right:
                break

            for i in range(right, left - 1, -1):
                ans.append(matrix[button][i])
            button -= 1
            if top > button:
                break

            for i in range(button, top - 1, -1):
                ans.append(matrix[i][left])
            left += 1
            if left > right:
                break

        return ans

File: 剑指offer/29.顺时针打印矩阵.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix)
    {
        int m = matrix.size();
        if (m == 0) {
            return {};
        }
        int n = matrix[0].size();

        int left = 0;
        int right = n;

        int top = 0;
        int button = m;

        vector<int> ans;
        while (left < right && top < button) {
            for (int i = left; i < right; i++) {
                ans.push_back(matrix[top][i]);
            }
            top++;

            for (int i = top; i < button; i++) {
                ans.push_back(matrix[i][right - 1]);
            }
            right--;

            if (left < right && top < button) {

                for (int i = right - 1; i >= left; i--) {
                    ans.push_back(matrix[button - 1][i]);
                }
                button--;

                for (int i = button - 1; i >= top; i--) {
                    ans.push_back(matrix[i][left]);
                }
                left++;
            }
        }

        return ans;
    }
};